pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -578,52 +579,63 @@ func (a *AmbientIndexImpl) handlePod(oldObj, newObj any, isDelete bool, c *Contr
 	if !isDelete {
 		wl = a.extractWorkload(p, c)
 	}
-	wlNetwork := c.Network(p.Status.PodIP, p.Labels).String()
-	networkAddr := networkAddress{network: wlNetwork, ip: p.Status.PodIP}
 	uid := c.generatePodUID(p)
 	oldWl := a.byUID[uid]
 	if wl == nil {
-		// This is an explicit delete event, or there is no longer a Workload to create (pod NotReady, etc)
-		delete(a.byPod, networkAddr)
-		delete(a.byUID, uid)
-		if oldWl != nil {
-			// If we already knew about this workload, we need to make sure we drop all service references as well
-			for namespacedHostname := range oldWl.Services {
-				a.dropWorkloadFromService(namespacedHostname, oldWl.ResourceName())
-			}
-			log.Debugf("%v: workload removed, pushing", p.Status.PodIP)
-			// TODO: namespace for network?
-			updates.Insert(model.ConfigKey{Kind: kind.Address, Name: oldWl.ResourceName()})
-			return updates
-		}
-		// It was a 'delete' for a resource we didn't know yet, no need to send an event
-
+		a.updateWorkloadIndexes(oldWl, nil, updates)
 		return updates
 	}
 	if oldWl != nil && proto.Equal(wl.Workload, oldWl.Workload) {
 		log.Debugf("%v: no change, skipping", wl.ResourceName())
-
 		return updates
 	}
-	for _, networkAddr := range networkAddressFromWorkload(wl) {
-		a.byPod[networkAddr] = wl
+	a.updateWorkloadIndexes(oldWl, wl, updates)
+
+	return updates
+}
+
+// updateWorkloadIndexes, given and old and new instance, updates the various indexes for workloads.
+// Any changes are reported in `updates`.
+func (a *AmbientIndexImpl) updateWorkloadIndexes(oldWl *model.WorkloadInfo, newWl *model.WorkloadInfo, updates sets.Set[model.ConfigKey]) {
+	if newWl == nil {
+		if oldWl == nil {
+			// No change needed
+			return
+		}
+		updates.Insert(model.ConfigKey{Kind: kind.Address, Name: oldWl.ResourceName()})
+		for _, addr := range networkAddressFromWorkload(oldWl) {
+			if oldWl.Source == model.WorkloadSourcePod {
```

Comment:
i noticed `WorkloadSourceServiceEntry` is not checked here (and below). is that ok?

Reply:
`byWorkloadEntry` stores entries from WE and from SE, so it is intended